external-crates/move/crates/move-compiler/src/typing/deprecations.rs

```diff
@@ -0,0 +1,765 @@
+                }
+            }
+
+            for (name, datatype) in module_def.enums.key_cloned_iter() {
+                if let Some(deprecation) = deprecations(
+                    s.env,
+                    AttributePosition::Enum,
+                    &datatype.attributes,
+                    name.loc(),
+                    mident,
+                ) {
+                    s.deprecated_types.insert((mident, name), deprecation);
+                }
+            }
+        }
+
+        s
+    }
+
+    // Build up a position index/position map for all the members in the program (including dependencies).
+    fn index(prog: &T::Program) -> BTreeMap<ModuleMemberRef<Name>, AttributePosition> {
+        let mut member_index = BTreeMap::new();
+        for (mident, module_info) in prog.info.modules.key_cloned_iter() {
+            for (name, _) in module_info.constants.key_cloned_iter() {
+                member_index.insert((mident, name.0), AttributePosition::Constant);
+            }
+            for (name, _) in module_info.functions.key_cloned_iter() {
+                member_index.insert((mident, name.0), AttributePosition::Function);
+            }
+            for (name, _) in module_info.structs.key_cloned_iter() {
+                member_index.insert((mident, name.0), AttributePosition::Struct);
+            }
+            for (name, _) in module_info.enums.key_cloned_iter() {
+                member_index.insert((mident, name.0), AttributePosition::Enum);
+            }
+        }
+
+        for (mident, module_def) in prog.modules.key_cloned_iter() {
+            for (name, _) in module_def.constants.key_cloned_iter() {
+                member_index.insert((mident, name.0), AttributePosition::Constant);
+            }
+
+            for (name, _) in module_def.functions.key_cloned_iter() {
+                member_index.insert((mident, name.0), AttributePosition::Function);
+            }
+
+            for (name, _) in module_def.structs.key_cloned_iter() {
+                member_index.insert((mident, name.0), AttributePosition::Struct);
+            }
+
+            for (name, _) in module_def.enums.key_cloned_iter() {
+                member_index.insert((mident, name.0), AttributePosition::Enum);
+            }
+        }
+
+        member_index
+    }
+
+    fn set_module(&mut self, mident: ModuleIdent) {
+        self.current_mident = Some(mident);
+    }
+
+    fn set_named_context(&mut self, named_context: Name) {
+        self.current_named_context = Some(named_context);
+    }
+
+    // Look up the deprecation information for a given type, constant, or function. If none is
+    // found, see if the module that the type is in is deprecated.
+    fn deprecated_type(
+        &self,
+        mident: &ModuleIdent,
+        name: &DatatypeName,
+    ) -> Option<(Deprecation, AttributePosition)> {
+        let position = *self
+            .member_index
```

Comment:
Looks like this is the only place this is read.
I would replace this with `TypingProgramInfo` 

The function you need _might_ not exist, but if it doesn't, I would add it.

Reply:
Great suggestion -- made it much nicer :)